src/v/raft/replicate_batcher.cc

```diff
@@ -281,14 +292,65 @@ ss::future<> replicate_batcher::do_flush(
   append_entries_request req,
   std::vector<ss::semaphore_units<>> u,
   absl::flat_hash_map<vnode, follower_req_seq> seqs) {
+    auto needs_flush = req.flush;
     _ptr->_probe.replicate_batch_flushed();
+    auto stm = ss::make_lw_shared<replicate_entries_stm>(
+      _ptr, std::move(req), std::move(seqs));
     try {
-        auto result = co_await _ptr->dispatch_replicate(
-          std::move(req), std::move(u), std::move(seqs));
+        auto holder = _bg.hold();
+        auto leader_result = co_await stm->append_to_leader(std::move(u));
 
-        propagate_result(result, notifications);
+        /**
+         * First phase, if leader result has error just propagate error
+         * otherwise propagate result to relaxed consistency requests
+         */
+        propagate_result(
+          leader_result,
+          notifications,
+          [has_error = leader_result.has_error()](const item_ptr& item) {
+              return has_error
+                     || item->consistency_lvl == consistency_level::leader_ack
+                     || item->consistency_lvl == consistency_level::no_ack;
+          });
+        /**
+         * Second phase, wait for majority to replicate if leader result has
+         * error just propagate error otherwise propagate results to relaxed
+         * consistency requests.
+         *
+         * NOTE: this happens in background since we do not want to block
+         * replicate batcher
+         */
+        if (leader_result && needs_flush) {
+            (void)stm->wait_for_majority()
+              .then([holder = std::move(holder),
+                     notifications = std::move(notifications)](
+                      result<replicate_result> quorum_result) mutable {
+                  propagate_result(
+                    quorum_result, notifications, [](const item_ptr& item) {
+                        return item->consistency_lvl
+                               == consistency_level::quorum_ack;
+                    });
+              })
+              .finally([stm] {});
+        }
     } catch (...) {
         propagate_current_exception(notifications);
     }
+
+    auto f = stm->wait().finally([stm] {});
+    if (_bg.is_closed()) {
+        _ptr->_ctxlog.info(
+          "gate-closed, waiting to finish background requests");
+        co_return co_await std::move(f);
+    }
+    // background
+    ssx::spawn_with_gate(_bg, [this, stm, f = std::move(f)]() mutable {
+        return std::move(f).handle_exception(
+          [this](const std::exception_ptr& e) {
+              _ptr->_ctxlog.error(
+                "Error waiting for background acks to finish - {}", e);
```

Comment:
Maybe this should be a lower severity than error -- I think it'll happen any time we transfer leadership between a request getting a quorum ack and its final follower acks arriving, right?  So not really an error case